lib/Frontend/Frontend.cpp

```diff
@@ -440,141 +549,150 @@ void CompilerInstance::performSema() {
+    performPCMacro(MainFile, PersistentState.getTopLevelContext());
+  }
+
+  // Playground transform knows to look out for PCMacro's changes and not
+  // to playground log them.
+  if (mainIsPrimary && !Context->hadError() &&
+      Invocation.getFrontendOptions().PlaygroundTransform)
+    performPlaygroundTransform(
+        MainFile, Invocation.getFrontendOptions().PlaygroundHighPerformance);
+  if (!mainIsPrimary) {
+    performNameBinding(MainFile);
+  }
+}
+
+void CompilerInstance::typeCheckEveryFileInMainModule(
+                                                      PersistentParserState &PersistentState,
+                                                      const OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  const auto &options = Invocation.getFrontendOptions();
+  for (auto File : MainModule->getFiles()) {
+    if (auto SF = dyn_cast<SourceFile>(File)) {
+      performTypeChecking(*SF, PersistentState.getTopLevelContext(),
+                          TypeCheckOptions, /*curElem*/ 0,
+                          options.WarnLongFunctionBodies,
+                          options.WarnLongExpressionTypeChecking,
+                          options.SolverExpressionTimeThreshold);
+    }
+  }
+}
+void CompilerInstance::typeCheckThePrimaryFile(
+                                               PersistentParserState &PersistentState,
+                                               const OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  assert (PrimarySourceFile != nullptr  &&  "No primary file?");
+  const auto &options = Invocation.getFrontendOptions();
+  performTypeChecking(*PrimarySourceFile, PersistentState.getTopLevelContext(),
+                      TypeCheckOptions, /*curElem*/ 0,
+                      options.WarnLongFunctionBodies,
+                      options.WarnLongExpressionTypeChecking,
+                      options.SolverExpressionTimeThreshold);
+}
+
+
+
+void CompilerInstance::typeCheckMainModule(
+    OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  SharedTimer timer("performSema-checkTypesWhileParsingMain-typeCheckMainModule");
   if (TypeCheckOptions & TypeCheckingFlags::DelayWholeModuleChecking) {
-    for (auto File : MainModule->getFiles())
-      if (auto SF = dyn_cast<SourceFile>(File))
-        performWholeModuleTypeChecking(*SF);
+    performWholeModuleTypeCheckingOnMainModule();
   }
+  finishTypeCheckingMainModule();
+}
+
+void CompilerInstance::performWholeModuleTypeCheckingOnMainModule() {
+  SharedTimer timer("performWholeModuleTypeCheckingOnMainModule");
+  for (auto File : MainModule->getFiles())
+    if (auto SF = dyn_cast<SourceFile>(File))
+      performWholeModuleTypeChecking(*SF);
+}
 
+void CompilerInstance::finishTypeCheckingMainModule() {
+  SharedTimer timer("finishTypeCheckingMainModule");
   for (auto File : MainModule->getFiles())
     if (auto SF = dyn_cast<SourceFile>(File))
       if (PrimaryBufferID == NO_SUCH_BUFFER || SF == PrimarySourceFile)
```

Comment:
Last time there was a loop like this you decided to factor it out. Maybe there should be a higher-order `forEachFileToEmitCodeFor` or something?

Reply:
Yes, this is nice. Take a look.